exploration/function-composition-part-1.md

```diff
@@ -0,0 +1,824 @@
+
+Interpretation 2 treats variables analogously to cells in a spreadsheet.
+In a cell of a spreadsheet, referring to a value by name
+creates a reference,
+rather than copying its value.
+Unlike cells in spreadsheets, variables in MessageFormat are immutable.
+However, like cells in spreadsheets, the definitions of variables
+can refer to other variables that are in scope.
+
+Consider a tabular rendering of example A3 (with a pseudo-spreadsheet syntax)
+(this would be the "formula" view):
+
+|    | A       |     B   | C                |  D              |   E              |
+|----|---------|---------|------------------|-----------------|------------------|
+| 1  | 1.00123 | 1.00    | (=B1, maxFrac=2) | =(A1, maxFrac=2)| =(D1, maxFrac=5) |
+
+with the following mappings between cells and variables:
+
+| A1 | n1 |
+|----|----|
+| B1 | n2 |
+|----|----|
+| C1 | x0 |
+|----|----|
+| D1 | x  |
+|----|----|
+| E1 | y  |
+
+And the "output" view
+
+|    | A       |     B   | C                |  D              |   E              |
+|----|---------|---------|------------------|-----------------|------------------|
+1    | 1.00123 | 1.00    | 1.00             | 1.00            | 1.00123          |
+
+In E1, the reference to D1 is a reference to the _value_ of A1,
+with added formatting options that can be extended or overridden.
+
+This is consistent with interpretation 2, in which `$x` is bound to a structure
+that contains the value of `$n1`.
+(This could be implemented by copying, since MessageFormat variables are immutable.)
+
+
+### Semantics: return values for functions
+
+Turning to the question of what a function returns,
+this at first glance is the same question as
+"what is a named value?",
+as both are "resolved values" according to the spec.
+But both interpretation 1 and interpretation 2 complicate that.
+
+Alternative 1: A function returns a "formatted value".
+This matches model 1, where formatted values
+are bound to names.
+
+Alternative 2: A function returns a composite value
+that conceptually pairs a base value (possibly the
+operand of the function, but possibly not; see Example B1)
+with options.
+This matches model 2.
+If we preserve the single usage of "resolved value"
+in the spec, this implies that the (base value, options)
+representation applies to all resolved values,
+not just those returned by functions.
+
+For more details, see the "model implementations" section.
+
+### Allowing different kinds of functions
+
+The syntax could also distinguish multiple kinds of functions,
+some of which are composable and some not.
+
+Or, the function registry could allow functions to be declared in this way,
+with incorrect uses of functions being resolution errors.
+
+(See "Different kinds of composition" under "Alternatives to consider".)
+
+### Summarizing use cases
+
+There seem to be two areas of ambiguity:
+
+* Are named values essentially strings with metadata,
+or are they structured? (Model 1 vs. Model 2)
```

Comment:
I'm not sure that I understand what the difference between these is, or what the ambiguity here is. Doesn't a "string with metadata" need to have an object wrapper of some sort providing slots for where to put the metadata? How is that different from being "structured"?

Reply:
You're right, this is vague. I tried to address this in [09aeff1](https://github.com/unicode-org/message-format-wg/pull/753/commits/09aeff1afa1916f4d74603ebb3cfae6a94bf204c), which runs the hazard of getting too implementation-specific, but I'm not sure how else to express it.